ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BatchDeleteHandler.java

```diff
@@ -0,0 +1,181 @@
+
+      // Extract request path
+      RequestPath requestPath = (RequestPath) restRequest.getArgs().get(InternalKeys.REQUEST_PATH);
+      List<CompletableFuture<Void>> deleteFutures = new ArrayList<>();
+
+      if (deleteRequest.getObjects().size() > maxBatchSize) {
+        LOGGER.error("Exceeded batch size");
+        throw new RestServiceException("Batch Size Exceeded", RestServiceErrorCode.BadRequest);
+      }
+      for (S3MessagePayload.S3BatchDeleteKey object : deleteRequest.getObjects()) {
+
+        // Construct the delete path for each object
+        String singleDeletePath = requestPath.getPathAfterPrefixes() + "/" + object.getKey();
+
+        // Create a new RequestPath for the delete operation
+        List<String> emptyList = new ArrayList<>();
+        RequestPath newRequestPath =
+            RequestPath.parse(singleDeletePath, restRequest.getArgs(), emptyList, requestPath.getClusterName());
+        WrappedRestRequest singleDeleteRequest = new WrappedRestRequest(restRequest);
+        singleDeleteRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
+
+        NoOpResponseChannel noOpResponseChannel = new NoOpResponseChannel();
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        // Handle the delete operation using the deleteBlobHandler
+        deleteBlobHandler.handle(singleDeleteRequest, noOpResponseChannel, (result, exception) -> {
+          // Call our custom onDeleteCompletion to track success/failure
+          onDeleteCompletion(exception, object.getKey());
+          future.complete(null);
+        });
+        deleteFutures.add(future);
+      }
+
+      CompletableFuture.allOf(deleteFutures.toArray(new CompletableFuture[0]))
+          .whenComplete((result, exception) -> {
+            try {
+              resp.setDeletedKeys(deleted);
+              resp.setErrors(errors);
+              xmlMapper.writeValue(outputStream, resp);
+              ReadableStreamChannel finalreadableStreamChannel =
+                  new ByteBufferReadableStreamChannel(ByteBuffer.wrap(outputStream.toByteArray()));
+              restResponseChannel.setHeader(Headers.CONTENT_LENGTH, readableStreamChannel.getSize());
+              restResponseChannel.setHeader(Headers.CONTENT_TYPE, XML_CONTENT_TYPE);
+              restResponseChannel.setStatus(ResponseStatus.Ok);
+              finalCallback.onCompletion(finalreadableStreamChannel, null);
+            } catch (IOException | RestServiceException e) {
+              LOGGER.error("Failed to complete", e);
+              finalCallback.onCompletion(null, e);
+            }
+          });
+    }, restRequest.getUri(), LOGGER, finalCallback);
+  }
+
+  /**
+   * This method is used to update the lists of deleted and errored objects.
+   * @param exception whether the delete was successful or not
+   * @param key the object key
+   */
+  public void onDeleteCompletion(Exception exception, String key) {
+    if (exception == null) {
+      deleted.add(key);
```

Comment:
Multiple threads are going to call this fn concurrently. Do we need a some lock or a concurrent data structure here ?